gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageClientReadChannel.java

```diff
@@ -0,0 +1,505 @@
+        resetContentChannel();
+      }
+    }
+  }
+
+  private void performPendingSeeks(long bytesToRead) throws IOException {
+    if (currentPosition == contentChannelPosition && contentReadChannel != null) {
+      // No seek is required
+      return;
+    }
+
+    logger.atFiner().log(
+        "Performing lazySeek from %s to %s position for '%s'",
+        contentChannelPosition, currentPosition, resourceId);
+    // record the variation in channelPosition
+    // used to auto-detect random access
+    long oldPosition = contentChannelPosition;
+    long seekDistance = currentPosition - contentChannelPosition;
+    // Check if inPlace seek is a viable option
+    // If not close existing channel and open new one.
+
+    if (contentReadChannel != null
+        && seekDistance > 0
+        && seekDistance <= readOptions.getInplaceSeekLimit()
+        && currentPosition < contentChannelEnd) {
+      skipInPlace(seekDistance);
+    } else {
+      closeContentChannel();
+    }
+    if (contentReadChannel == null) {
+      if (isRandomAccessPattern(oldPosition)) {
+        setRandomAccess();
+      }
+      contentReadChannel = openReadChannel(bytesToRead);
+    }
+  }
+
+  // Update this function to create the reader and then set limits
+  private ReadableByteChannel getStorageReadChannel(long seek, long limit) throws IOException {
+    ReadChannel readChannel = storage.reader(blobId, generateReadOptions(blobId));
+    try {
+      readChannel.seek(seek);
+      // TODO: should we also add a check to not set limit if contentChannelEnd == objectSize?
+      if (limit != -1) {
+        readChannel.limit(limit);
+      }
+      return readChannel;
+    } catch (Exception e) {
+      throw new IOException(
+          String.format(
+              "Unable to update the boundaries/Range of contentChannel %s", resourceId.toString()),
+          e);
+    }
+  }
+
+  private static BlobSourceOption[] generateReadOptions(BlobId blobId) {
+    List<BlobSourceOption> readOptions = new ArrayList<>();
+    if (blobId.getGeneration() != null) {
+      readOptions.add(BlobSourceOption.generationMatch(blobId.getGeneration()));
+    }
+    return readOptions.toArray(new BlobSourceOption[readOptions.size()]);
+  }
+
+  private boolean isFooterRead() {
+    return objectSize - currentPosition <= footerSize;
+  }
+
+  private static void validate(GoogleCloudStorageItemInfo itemInfo) throws IOException {
+    checkNotNull(itemInfo, "itemInfo cannot be null");
+    StorageResourceId resourceId = itemInfo.getResourceId();
+    checkArgument(
+        resourceId.isStorageObject(), "Can not open a non-file object for read: %s", resourceId);
+    if (!itemInfo.exists()) {
+      throw new FileNotFoundException(String.format("File not found: %s", resourceId));
+    }
+    // The non-gRPC read channel has special support for gzip. This channel doesn't
+    // decompress gzip-encoded objects on the fly, so best to fail fast rather than return
```

Comment:
isn't this a change in behaviour?

Reply:
Not form gRPC point. We were not supporting `gzip` encoded objects via current GAPIC implementation too.

I have specifically not added support for it as of now because there is been some development going on in java-storage in `gzip` space which will leverage later on (before our release).

Reply:
Can we call out the limitations of using the GCS client as opposed to JSON client? It is not clear to me. I was of the impression that all the feature in JSON is supported in GCS Java client.